tests/src/lib.rs

```diff
@@ -53,11 +53,80 @@ fn lowest_downlevel_properties() -> DownlevelCapabilities {
     }
 }
 
-pub struct FailureCase {
-    backends: Option<wgpu::Backends>,
-    vendor: Option<u32>,
-    adapter: Option<String>,
-    skip: bool,
+/// Conditions under which a test should fail.
+///
+/// If any of these fields are `None`, then the field does not
+/// restrict matches. For example:
+///
+/// ```
+/// # use wgpu_test::FailureCase;
+/// FailureCase {
+///     backends: Some(wgpu::Backends::DX11 | wgpu::Backends::DX12),
+///     vendor: None,
+///     adapter: Some("RTX"),
+///     driver: None,
+///     skip: false,
+/// }
+/// # ;
+/// ```
+///
+/// This applies to all cards with `"RTX'` in their name on either
+/// Direct3D backend, no matter the vendor ID or driver name.
+///
+/// The default value of `FailureCase` applies to any test case (that
+/// is, there are no criteria to constrain the match), and the test is
+/// not skipped.
+pub struct FailureCase<S> {
+    /// Backends expected to fail, or `None` for any backend.
+    ///
+    /// If this is `None`, or if the test is using one of the backends
+    /// in `backends`, then this `FailureCase` applies.
+    pub backends: Option<wgpu::Backends>,
+
+    /// Vendor expected to fail, or `None` for any vendor.
+    ///
+    /// If `Some`, this must match [`AdapterInfo::device`], which is
+    /// usually the PCI device id. Otherwise, this `FailureCase`
+    /// applies regardless of vendor.
+    ///
+    /// [`AdapterInfo::device`]: wgt::AdapterInfo::device
+    pub vendor: Option<u32>,
+
+    /// Name of adaper expected to fail, or `None` for any adapter name.
+    ///
+    /// If this is `Some(s)` and `s` is a substring of
+    /// [`AdapterInfo::name`], then this `FailureCase` applies. If
+    /// this is `None`, the adapter name isn't considered.
+    ///
+    /// [`AdapterInfo::name`]: wgt::AdapterInfo::name
+    pub adapter: Option<S>,
+
+    /// Name of driver expected to fail, or `None` for any driver name.
+    ///
+    /// If this is `Some(s)` and `s` is a substring of
+    /// [`AdapterInfo::driver`], then this `FailureCase` applies. If
+    /// this is `None`, the driver name isn't considered.
+    ///
+    /// [`AdapterInfo::driver`]: wgt::AdapterInfo::driver
+    pub driver: Option<S>,
+
+    /// If `true`, skip the test altogether.
+    ///
+    /// If `false`, the test is run, and an unexpected pass is
+    /// reported if it doesn't fail.
+    pub skip: bool,
```

Comment:
It would improve readability signifigantly to have this be an enum between "ExpectFailure" and "skip" or something like that, I've already gotten tripped up reading the bools in this MR.

`.molten_vk_failure(false)` sounds like it shouldn't fail on moltenvk, but molten_vk_failure(FailureMode::ExpectFailure) or whatever is much better